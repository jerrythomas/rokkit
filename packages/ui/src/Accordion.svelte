<script>
	import { equals } from 'ramda'
	import { createEmitter, noop, getKeyFromPath, getSnippet } from '@rokkit/core'
	import { NestedController } from '@rokkit/states'
	import { navigator } from '@rokkit/actions'
	import Summary from './Summary.svelte'
	import Item from './Item.svelte'
	import ListBody from './ListBody.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {import('@rokkit/core').FieldMapper} [mapping]
	 * @property {boolean} [autoCloseSiblings]
	 * @property {boolean} [multiselect]
	 * @property {any} [value]
	 */

	/** @type {Props} */
	let {
		class: classes = '',
		items = $bindable([]),
		value = $bindable(null),
		fields,
		autoCloseSiblings = false,
		multiselect = false,
		header = null,
		footer = null,
		empty = null,
		stub = null,
		extra,
		...events
	} = $props()

	let emitter = $derived(
		createEmitter(events, ['collapse', 'change', 'expand', 'click', 'select', 'move'])
	)
	let wrapper = new NestedController(items, value, fields, {
		events,
		multiselect,
		autoCloseSiblings
	})
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<rk-accordion
	class={classes}
	tabindex="0"
	use:navigator={{ wrapper }}
	onactivate={() => (value = wrapper.value)}
>
	{#if header}
		<rk-header>{@render header()}</rk-header>
	{/if}
	{#if items.length === 0}
		<rk-list-item role="presentation">
			{#if empty}
				{@render empty()}
			{:else}
				No items found.
			{/if}
		</rk-list-item>
	{/if}
	{#each items as item, index}
		{@const key = `${index}`}
		<div
			class="flex flex-col"
			class:is-expanded={wrapper.expandedKeys.has(key)}
			class:is-selected={wrapper.selectedKeys.has(key)}
			data-path={index}
		>
			<Summary
				bind:value={items[index]}
				{fields}
				expanded={wrapper.expandedKeys.has(key)}
				hasChildren={wrapper.hasChildren(key)}
			/>
			{#if wrapper.expandedKeys.has(key)}
				<rk-list role="listbox" tabindex="-1">
					<ListBody
						bind:items={items[fields.children]}
						bind:value
						fields={fields.fields ?? fields}
						{selected}
						{stub}
						onchange={emitter.change}
						{extra}
					/>
				</rk-list>
			{/if}
		</div>
	{/each}
	{#if footer}
		<rk-footer>{@render footer()}</rk-footer>
	{/if}
</rk-accordion>
