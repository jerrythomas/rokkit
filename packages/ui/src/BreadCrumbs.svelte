<script>
	import { getSnippet } from '@rokkit/core'
	import Item from './Item.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [items]
	 * @property {string} [separator]
	 * @property {any} [fields]
	 * @property {any} [crumb]
	 */

	/** @type {Props} */
	let { class: classes = '', items = [], separator = '/', fields, crumb } = $props()
</script>

<rk-crumbs class={classes}>
	{#each items as item, index (index)}
		{#if index > 0}
			<span>
				{#if separator.length === 1}
					{separator}
				{:else}
					<icon class={separator}></icon>
				{/if}
			</span>
		{/if}
		<rk-crumb class:is-selected={index === items.length - 1}>
			{#if crumb}
				{@render crumb(item, fields)}
			{:else}
				<Item value={item} {fields} />
			{/if}
		</rk-crumb>
	{/each}
</rk-crumbs>
