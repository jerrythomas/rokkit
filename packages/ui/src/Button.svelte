<script>
	/**
	 * @typedef {Object} ButtonProps
	 * @property {'default' | 'primary' | 'secondary' | 'tertiary'} variant - The variant of the button.
	 * @property {'button' | 'submit' | 'reset'} type - The type of the button.
	 * @property {string} [label] - The label of the button.
	 * @property {string|import('svelte').Snippet} [leftIcon] - The left icon of the button.
	 * @property {string|import('svelte').Snippet} [rightIcon] - The right icon of the button.
	 * @property {import('svelte').Snippet} [children]
	 * @property {Function} onclick - The onclick event of the button.
	 * @property {Function} onsubmit - The onsubmit event of the button.
	 * @property {Function} onreset - The onreset event of the button.
	 */
	/** @type {ButtonProps} */
	let {
		variant = 'default',
		type = 'button',
		leftIcon = null,
		rightIcon = null,
		label = null,
		children = null,
		description = null,
		onclick,
		onsubmit,
		onreset
	} = $props()

	const primary = $derived(variant === 'primary')
	const secondary = $derived(variant === 'secondary')
	const tertiary = $derived(variant === 'tertiary')
</script>

<button
	class:primary
	class:secondary
	class:tertiary
	class={classes}
	{type}
	{onclick}
	{onsubmit}
	{onreset}
	aria-label={description ?? label}
>
	{#if typeof leftIcon === 'string'}
		<Icon name={leftIcon} />
	{:else}
		{@render leftIcon?.()}
	{/if}

	{#if children}
		{@render children()}
	{:else if label}
		{label}
	{/if}

	{#if typeof rightIcon === 'string'}
		<Icon name={rightIcon} />
	{:else}
		{@render rightIcon?.()}
	{/if}
</button>
