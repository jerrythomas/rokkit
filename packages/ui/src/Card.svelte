<script>
	import { Proxy } from '@rokkit/states'
	let { class: classNames = '', value = $bindable(), fields, child, onClick } = $props()

	const proxy = $state(new Proxy(value, fields))
	const childSnippet = $derived(child ?? defaultChild)
	const href = $derived(proxy.get('href'))

	/**
	 * Handles enter key
	 * @param {KeyboardEvent} event
	 */
	function handleKeyUp(event) {
		if (event.key === 'Enter') onClick?.()
	}
</script>

{#snippet defaultChild(/** @type {Proxy}*/ proxy)}
	<icon class="flex flex-col">
		<i class={proxy.get('icon')}></i>
	</icon>
	<h1>{proxy.get('title')}</h1>
	<h2>{proxy.get('description')}</h2>
{/snippet}

{#if href}
	<a {href} data-card-root class={classNames}>
		{@render childSnippet(proxy)}
	</a>
{:else}
	<div
		data-card-root
		onclick={onClick}
		onkeyup={handleKeyUp}
		role="button"
		tabindex="-1"
		class={classNames}
	>
		{@render childSnippet(proxy)}
	</div>
{/if}
