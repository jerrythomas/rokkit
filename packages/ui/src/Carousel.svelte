<script>
	let className = ''
	export { className as class }
	export let items = []
	export let value
	let currentIndex = 0

	function handleKey(e, index) {
		const prevIndex = currentIndex
		if (['Enter', ' '].includes(e.key)) {
			currentIndex = index
		} else {
			if (e.key === 'ArrowRight') {
				currentIndex = currentIndex === items.length - 1 ? 0 : currentIndex + 1
			} else if (e.key === 'ArrowLeft') {
				currentIndex = currentIndex === 0 ? items.length - 1 : currentIndex - 1
			}
		}
		if (currentIndex !== prevIndex) {
			e.preventDefault()
			e.stopPropagation()
		}
	}
	$: value = items[currentIndex]
</script>

{#if items.length === 0}
	<p>No items to display</p>
{:else}
	<carousel class="flex w-full flex-col gap-3 {className}">
		<slide class="flex flex-col">
			<img src={value.src} alt={value.alt} />
			<p>{value.text}</p>
		</slide>
		<dot-nav role="radiogroup">
			{#each items as item, index (index)}
				<dot
					role="radio"
					aria-checked={currentIndex === index}
					aria-label={`Slide ${index + 1}`}
					on:click={() => (currentIndex = index)}
					on:keydown={(e) => handleKey(e, index)}
					tabindex={0}
				>
				</dot>
			{/each}
		</dot-nav>
	</carousel>
{/if}
