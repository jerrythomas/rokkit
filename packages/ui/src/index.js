// skipcq: JS-E1004 - Needed for exposing all components
export * from '@rokkit/atoms'
// skipcq: JS-E1004 - Needed for exposing all components
export * from '@rokkit/input'
