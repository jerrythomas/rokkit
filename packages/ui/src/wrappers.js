// skipcq: JS-E1004 - Needed for exposing all components
export * from '@rokkit/organisms/wrappers'
