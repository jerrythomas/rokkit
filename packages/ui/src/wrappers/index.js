export { default as Wrapper } from './Wrapper.svelte'
export { default as Section } from './Section.svelte'
export { default as Category } from './Category.svelte'
