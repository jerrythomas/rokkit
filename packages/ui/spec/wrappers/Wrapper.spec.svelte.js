import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { flushSync, tick } from 'svelte'
import { Wrapper } from '../../src/wrappers'

describe('Wrapper.svelte', () => {
	beforeEach(() => cleanup())

	it('should render using default props', () => {
		const { container } = render(Wrapper)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render using specified type', async () => {
		const props = $state({ type: 'horizontal' })
		const { container } = render(Wrapper, { props })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		// handle type changes
		props.type = 'section'
		flushSync()
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render using specified class', async () => {
		const props = $state({ class: 'custom-class' })
		const { container } = render(Wrapper, { props })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		props.class = 'custom-class-2'
		flushSync()
		await tick()
		expect(container).toMatchSnapshot()
	})
})
