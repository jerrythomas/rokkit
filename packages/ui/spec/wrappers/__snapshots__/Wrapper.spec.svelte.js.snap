// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Wrapper.svelte > should render using default props 1`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical"
  >
    <!---->
  </field-layout>
  
</div>
`;

exports[`Wrapper.svelte > should render using specified class 1`] = `
<div>
  <field-layout
    aria-label="original"
    class="custom-class vertical"
  >
    <!---->
  </field-layout>
  
</div>
`;

exports[`Wrapper.svelte > should render using specified class 2`] = `
<div>
  <field-layout
    aria-label="new label"
    class="custom-class-2 vertical"
  >
    <!---->
  </field-layout>
  
</div>
`;

exports[`Wrapper.svelte > should render using specified type 1`] = `
<div>
  <field-layout
    aria-label=""
    class="horizontal"
  >
    <!---->
  </field-layout>
  
</div>
`;

exports[`Wrapper.svelte > should render using specified type 2`] = `
<div>
  <field-layout
    aria-label=""
    class="vertical"
  >
    <!---->
  </field-layout>
  
</div>
`;
